NorthstarDedicatedTest/sv_rcon.cpp

```diff
@@ -0,0 +1,478 @@
+	return false;
+}
+
+//-----------------------------------------------------------------------------
+// Purpose: handles input command buffer
+// Input  : *pszIn -
+//			nRecvLen -
+//			*pData -
+//-----------------------------------------------------------------------------
+void CRConServer::ProcessBuffer(const char* pszIn, int nRecvLen, CConnectedNetConsoleData* pData)
+{
+	while (nRecvLen)
+	{
+		switch (*pszIn)
+		{
+		case '\r':
+		{
+			if (pData->m_nCharsInCommandBuffer)
+			{
+				cl_rcon::request cl_request = this->Deserialize(pData->m_pszInputCommandBuffer);
+				this->ProcessMessage(cl_request);
+			}
+			pData->m_nCharsInCommandBuffer = 0;
+			break;
+		}
+		default:
+		{
+			if (pData->m_nCharsInCommandBuffer < MAX_NETCONSOLE_INPUT_LEN - 1)
+			{
+				pData->m_pszInputCommandBuffer[pData->m_nCharsInCommandBuffer++] = *pszIn;
+			}
+			break;
+		}
+		}
+		pszIn++;
+		nRecvLen--;
+	}
+}
+
+//-----------------------------------------------------------------------------
+// Purpose: processes received message
+// Input  : *cl_request -
+//-----------------------------------------------------------------------------
+void CRConServer::ProcessMessage(const cl_rcon::request& cl_request)
+{
+	CConnectedNetConsoleData* pData = m_pSocket->GetAcceptedSocketData(m_nConnIndex);
+
+	if (!pData->m_bAuthorized && cl_request.requesttype() != cl_rcon::request_t::SERVERDATA_REQUEST_AUTH)
+	{
+		// Notify net console that authentication is required.
+		std::string svMessage = this->Serialize(s_pszNoAuthMessage, "", sv_rcon::response_t::SERVERDATA_RESPONSE_AUTH);
+		::send(pData->m_hSocket, svMessage.c_str(), static_cast<int>(svMessage.size()), MSG_NOSIGNAL);
+
+		pData->m_bValidated = false;
+		pData->m_nIgnoredMessage++;
+		return;
+	}
+	switch (cl_request.requesttype())
+	{
+	case cl_rcon::request_t::SERVERDATA_REQUEST_AUTH:
+	{
+		this->Authenticate(cl_request, pData);
+		break;
+	}
+	case cl_rcon::request_t::SERVERDATA_REQUEST_EXECCOMMAND:
+	case cl_rcon::request_t::SERVERDATA_REQUEST_SETVALUE:
+	{
+		// Only execute if auth was succesfull.
+		if (pData->m_bAuthorized)
+		{
+			this->Execute(cl_request);
+		}
+		break;
+	}
+	case cl_rcon::request_t::SERVERDATA_REQUEST_SEND_CONSOLE_LOG:
+	{
+		if (pData->m_bAuthorized)
+		{
+			CVar_sv_rcon_sendlogs->SetValue(1);
+		}
+		break;
+	}
+	default:
+	{
+		break;
+	}
+	}
+}
+
+//-----------------------------------------------------------------------------
+// Purpose: execute commands issued from net console
+// Input  : *cl_request -
+//-----------------------------------------------------------------------------
+void CRConServer::Execute(const cl_rcon::request& cl_request) const
+{
+	ConVar* pConVar = g_pCVar->FindVar(cl_request.requestbuf().c_str());
+	if (pConVar)
+	{
+		pConVar->SetValue(cl_request.requestval().c_str());
+	}
+	else // Execute command with "<val>".
+	{
+		Cbuf_AddText(Cbuf_GetCurrentPlayer(), cl_request.requestbuf().c_str(), cmd_source_t::kCommandSrcCode);
+		Cbuf_Execute();
+	}
```

Comment:
Since both `SETVALUE` and `EXECCOMMAND` can go here, this could be dangerous:

If I hypothetically send a `SETVALUE` request with `requestBuf="quit"` and `requestVal="potatos"`, there is likely no convar called `quit` so instead this will execute that as a command. 

I don't believe it should ever be possible to execute a command with a `SETVALUE` request.

Reply:
There is no danger or undefined behavior here. If you run 'quit "potatos"' or 'exit 255 255 0 255', the engine would still call the 'quit' or 'exit' command callback which is just setting the statemachine's next state to HS_SHUTDOWN. The engine internally checks through the ConCommandBase class whether or not this is a ConVar through virtual method 'ConCommandBase::IsCommand()' (ConVar's and ConCommand's inherit this class for this reason, since they share a good part of the same system), if its a ConCommand it will run it as a such, even if you feed it 500 extra params and values since pCommandBase->IsCommand() returned true.

Reply:
Ah sorry, I didn’t mean to imply there’s undefined behaviour. What I mean is:

If I send a `SETVALUE` request with the variable set to `quit`, I would expect this to do nothing if that convar doesn’t exist - since I’m sending a `SETVALUE` and not an `EXECCOMMAND` request. Instead, as written it will fall back to running the variable as a command when it doesn’t exist, which seems counterintuitive. Hypothetical of course and it would be weird if someone tried to set the value of `quit`, but this could work for any other command too.

It’s not dangerous as in undefined behaviour, but as in it will run a command when you might not expect it to.

Would it make sense to handle SETVALUE and EXECCOMMAND separately?

Reply:
Yes, this makes sense. I changed the behavior accordingly.
The SETVALUE request will now only work on ConVar's as of commit 566d3dc33b9c8be39efe429f8ca848f3ff7580ea.